### PR TITLE
Rearrange FOP->create

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Needs https://github.com/OP-TEE/optee_test/pull/199 for tests to pass.

### CODE REVIEW

File: core/tee/tee_svc_storage.c [hunk 1]
```diff
@@ -432,6 +350,68 @@ TEE_Result syscall_storage_obj_open(unsigned long storage_id, void *object_id,
...
+	head.maxKeySize = o->info.maxKeySize;
+	head.objectUsage = o->info.objectUsage;
+	head.objectType = o->info.objectType;
+	head.have_attrs = o->have_attrs;
+
+	res = fops->create(o->pobj, !!(o->flags & TEE_DATA_FLAG_OVERWRITE),
+			   &head, sizeof(head), attr, attr_size, data, len,
+			   &o->fh);
+
+	if (!res)
```
Comment (etienne-lms) [inline core/tee/tee_svc_storage.c]: `res != TEE_SUCCESS`

File: core/tee/tee_sql_fs.c [hunk 2]
```diff
@@ -491,14 +428,161 @@ static TEE_Result sql_fs_write(struct tee_file_handle *fh, size_t pos,
...
+	}
+
+out:
+	if (res)
+		po->temporary = old_temporary;
+	sql_fs_end_transaction_rpc(res != TEE_SUCCESS);
+	mutex_unlock(&sql_fs_mutex);
+
+	return res;
+
```
Comment (etienne-lms) [inline core/tee/tee_sql_fs.c]: minor: empty line

File: core/tee/tee_sql_fs.c [hunk 3]
```diff
@@ -491,14 +428,161 @@ static TEE_Result sql_fs_write(struct tee_file_handle *fh, size_t pos,
...
+	TEE_Result res;
+	size_t pos = 0;
+	bool old_temporary;
+
+	mutex_lock(&sql_fs_mutex);
+	sql_fs_begin_transaction_rpc();
+	old_temporary = po->temporary;
+
+	res = open_internal(po, true, fh);
+	if (res)
```
Comment (etienne-lms) [inline core/tee/tee_sql_fs.c]: Here and 'till the end of the function: `if (res != TEE_SUCCESS)`

File: core/tee/tee_ree_fs.c [hunk 4]
```diff
@@ -424,32 +365,157 @@ static TEE_Result ree_fs_write(struct tee_file_handle *fh, size_t pos,
...
+	mutex_lock(&ree_fs_mutex);
+
+	res = open_internal(po, true, fh);
+	if (res)
+		goto out;
+
+	if (head && head_size) {
+		res = ree_fs_write_primitive(*fh, pos, head, head_size);
+		if (res)
+			goto out;
```
Comment (jenswi-linaro) [inline core/tee/tee_ree_fs.c]: Thanks, I'll fix.

File: core/tee/tee_rpmb_fs.c [hunk 5]
```diff
@@ -2518,13 +2543,74 @@ static void rpmb_fs_closedir(struct tee_fs_dir *dir)
...
+	bool old_temporary;
+
+	mutex_lock(&rpmb_mutex);
+	old_temporary = po->temporary;
+	res = rpmb_fs_open_internal(po, true, fh);
+
+	if (head && head_size) {
+		res = rpmb_fs_write_primitive(*fh, pos, head, head_size);
+		if (res)
+			goto out;
```
Comment (jenswi-linaro) [inline core/tee/tee_rpmb_fs.c]: Thanks, I'll fix.

File: core/tee/tee_sql_fs.c [hunk 6]
```diff
@@ -491,14 +428,161 @@ static TEE_Result sql_fs_write(struct tee_file_handle *fh, size_t pos,
...
+		goto out;
+
+	if (head && head_size) {
+		res = sql_fs_write_primitive(*fh, pos, head, head_size);
+		if (res)
+			goto out;
```
Comment (jenswi-linaro) [inline core/tee/tee_sql_fs.c]: Thanks, closing is enough here as `sql_fs_end_transaction_rpc(res != TEE_SUCCESS);` will roll everything back on failure.

**Comment by jenswi-linaro** [comment]
Addressed all comments except the `TEE_SUCCESS` related ones.

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by jenswi-linaro** [comment]
Tag applied, rebased.

**Pull request merged**